pkg/server/sandbox_stop.go

```diff
@@ -83,69 +80,33 @@ func (c *criContainerdService) StopPodSandbox(ctx context.Context, r *runtime.St
 		return nil, fmt.Errorf("failed to unmount sandbox files in %q: %v", sandboxRoot, err)
 	}
 
-	if err := c.stopSandboxContainer(ctx, id); err != nil {
+	if err := c.stopSandboxContainer(ctx, sandbox.Container); err != nil {
 		return nil, fmt.Errorf("failed to stop sandbox container %q: %v", id, err)
 	}
 	return &runtime.StopPodSandboxResponse{}, nil
 }
 
 // stopSandboxContainer kills and deletes sandbox container.
-func (c *criContainerdService) stopSandboxContainer(ctx context.Context, id string) error {
-	cancellable, cancel := context.WithCancel(ctx)
-	eventstream, err := c.eventService.Subscribe(cancellable, &events.SubscribeRequest{})
+func (c *criContainerdService) stopSandboxContainer(ctx context.Context, container containerd.Container) error {
+	task, err := container.Task(ctx, nil)
 	if err != nil {
-		return fmt.Errorf("failed to get containerd event: %v", err)
-	}
-	defer cancel()
-
-	resp, err := c.taskService.Get(ctx, &tasks.GetTaskRequest{ContainerID: id})
-	if err != nil {
-		if isContainerdGRPCNotFoundError(err) {
+		if errdefs.IsNotFound(err) {
 			return nil
 		}
 		return fmt.Errorf("failed to get sandbox container: %v", err)
 	}
-	if resp.Task.Status != task.StatusStopped {
-		// TODO(random-liu): [P1] Handle sandbox container graceful deletion.
-		if _, err := c.taskService.Kill(ctx, &tasks.KillRequest{
-			ContainerID: id,
-			Signal:      uint32(unix.SIGKILL),
-			All:         true,
-		}); err != nil && !isContainerdGRPCNotFoundError(err) && !isRuncProcessAlreadyFinishedError(err) {
-			return fmt.Errorf("failed to kill sandbox container: %v", err)
-		}
 
-		if err := c.waitSandboxContainer(eventstream, id, resp.Task.Pid); err != nil {
-			return fmt.Errorf("failed to wait for pod sandbox to stop: %v", err)
-		}
-	}
+	taskStatus, err := task.Status(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get task status for sandbox %q : %v", container.ID(), err)
 
-	// Delete the sandbox container from containerd.
-	_, err = c.taskService.Delete(ctx, &tasks.DeleteTaskRequest{ContainerID: id})
-	if err != nil && !isContainerdGRPCNotFoundError(err) {
-		return fmt.Errorf("failed to delete sandbox container: %v", err)
 	}
-	return nil
-}
-
-// waitSandboxContainer wait sandbox container stop event.
-func (c *criContainerdService) waitSandboxContainer(eventstream events.Events_SubscribeClient, id string, pid uint32) error {
-	for {
-		evt, err := eventstream.Recv()
-		if err != nil {
-			return err
-		}
-		// Continue until the event received is of type task exit.
-		if !typeurl.Is(evt.Event, &events.TaskExit{}) {
-			continue
-		}
-		any, err := typeurl.UnmarshalAny(evt.Event)
-		if err != nil {
-			return err
-		}
-		e := any.(*events.TaskExit)
-		if e.ContainerID == id && e.Pid == pid {
-			return nil
+	if taskStatus.Status != containerd.Stopped {
```

Comment:
Since we have `WithProcessKill` now, just `Delete`, we don't need check the status anymore.

In your current code, we'll never delete the task once the task is stopped.